### PR TITLE
Point to cause of `fn` expected return type

### LINKED ISSUE #48136: Confusing mismatched type error (part 2)
Related to (but different from) #43608.

If you have a function whose return type uses impl trait, you can get confusing type inference errors. The following code...

```rust
fn foo() -> impl std::fmt::Debug {
    if false {
        return;
    }
    
    "hello"
}
```

...produces the following error:

```text
error[E0308]: mismatched types
  --> src/main.rs:12:5
   |
12 |     "hello"
   |     ^^^^^^^ expected (), found reference
   |
   = note: expected type `()`
              found type `&'static str`
```

In a much more complicated and less obvious instance of this issue, I spent a long time trying to figure out why the compiler thought that `impl std::fmt::Debug` implied that my return value had to be `()`. It would be very helpful if the compiler could say something like "expected type `()` because of this line" and point to the original `return` (or whatever line caused the inference).

Without impl trait, the issue is less prominent because the types must match exactly, and as a result, if the two different return values have different types, at least one of them must fail to match the stated return type in the function declaration. Consider, for example, the following code:

```rust
fn bar() -> &'static str {
    if false {
        return;
    }
    
    "hello"
}
```

which gives the following error:

```text
error[E0069]: `return;` in a function whose return type is not `()`
  --> src/main.rs:17:9
   |
17 |         return;
   |         ^^^^^^ return type is not ()
```

The issue with impl trait is that both return values might implement the trait (in this case, both `()` and `&'static str` implement `Debug`), and so you don't get the same useful messages that you would with a concrete return type.

### PR DESCRIPTION
**Pull request by estebank**
Fix #48136.

### CODE REVIEW

File: src/librustc_typeck/check/coercion.rs [hunk 1]
```diff
@@ -1184,13 +1183,18 @@ impl<'gcx, 'tcx, 'exprs, E> CoerceMany<'gcx, 'tcx, 'exprs, E>
                     (self.final_ty.unwrap_or(self.expected_ty), expression_ty)
                 };
 
+                let reason_label = if label_expression_as_expected {
+                    "found because of this statement"
```
Comment (zackmdavis) [inline src/librustc_typeck/check/coercion.rs]: It doesn't look like the "found because ..." label gets exercised in the UI tests?

Comment (estebank) [reply src/librustc_typeck/check/coercion.rs]: I think that might be a dead code branch. Originally I needed a way to correct the label for a case where we had

```
fn foo() -> impl Debug {
    (return 1, return 2)
}
```

and had the expected/found order reversed, but after adding the overlapping check in 1224 this check is no longer needed and can be removed.

**Pull request merged**